src/3d/qgs3dsceneexporter.cpp

```diff
@@ -0,0 +1,804 @@
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgs3dsceneexporter.h"
+
+#include <QVector>
+#include <Qt3DCore/QEntity>
+#include <Qt3DCore/QComponent>
+#include <Qt3DCore/QNode>
+#include <Qt3DRender/QGeometry>
+#include <Qt3DRender/QAttribute>
+#include <Qt3DRender/QBuffer>
+#include <Qt3DRender/QGeometryRenderer>
+#include <Qt3DExtras/QPlaneGeometry>
+#include <Qt3DCore/QTransform>
+#include <Qt3DRender/QMaterial>
+#include <Qt3DExtras/QDiffuseMapMaterial>
+#include <Qt3DExtras/QTextureMaterial>
+#include <Qt3DRender/QTextureImage>
+#include <Qt3DRender/QTexture>
+#include <Qt3DRender/QBufferDataGenerator>
+#include <Qt3DRender/QBufferDataGeneratorPtr>
+#include <Qt3DRender/QMesh>
+#include <Qt3DRender/QSceneLoader>
+#include <Qt3DRender/QAbstractTexture>
+#include <Qt3DExtras/QCylinderGeometry>
+#include <Qt3DExtras/QConeGeometry>
+#include <Qt3DExtras/QSphereGeometry>
+#include <Qt3DExtras/QCuboidGeometry>
+#include <Qt3DExtras/QTorusGeometry>
+#include <Qt3DExtras/QExtrudedTextMesh>
+#include <Qt3DExtras/QPhongMaterial>
+#include <Qt3DRender/QAbstractTextureImage>
+
+#include <QByteArray>
+#include <QFile>
+#include <QTextStream>
+
+#include "qgstessellatedpolygongeometry.h"
+#include "qgsterraintileentity_p.h"
+#include "qgsterrainentity_p.h"
+#include "qgschunknode_p.h"
+#include "qgsterraingenerator.h"
+#include "qgs3dmapsettings.h"
+#include "qgsflatterraingenerator.h"
+#include "qgsdemterraingenerator.h"
+#include "qgsdemterraintileloader_p.h"
+#include "qgsdemterraintilegeometry_p.h"
+#include "qgs3dexportobject.h"
+#include "qgsterraintextureimage_p.h"
+#include "qgsterraintexturegenerator_p.h"
+#include "qgsmesh3dgeometry_p.h"
+#include "qgsmeshlayer.h"
+#include "qgsmesh3dentity_p.h"
+#include "qgsmeshterraingenerator.h"
+#include "qgsvectorlayer.h"
+#include "qgsabstract3drenderer.h"
+#include "qgsabstractvectorlayer3drenderer.h"
+#include "qgsvectorlayer3drenderer.h"
+#include "qgspolygon3dsymbol.h"
+#include "qgsline3dsymbol.h"
+#include "qgspoint3dsymbol.h"
+#include "qgsrulebased3drenderer.h"
+#include "qgs3dutils.h"
+#include "qgsbillboardgeometry.h"
+#include "qgsimagetexture.h"
+
+#include <numeric>
+
+template<typename T>
+QVector<T> getAttributeData( Qt3DRender::QAttribute *attribute, QByteArray data )
+{
+  uint bytesOffset = attribute->byteOffset();
+  uint bytesStride = attribute->byteStride();
+  uint vertexSize = attribute->vertexSize();
+  QVector<T> result;
+
+  if ( bytesStride == 0 )
+  {
+    qDebug() << "WARNING: bytesStride==0 at " << __FILE__ << ":" << __LINE__;
+    return result;
+  }
+
+  for ( int i = bytesOffset; i < data.size(); i += bytesStride )
+  {
+    for ( unsigned int j = 0; j < vertexSize * sizeof( T ); j += sizeof( T ) )
+    {
+      // maybe a problem with indienness can happen?
```

Comment:
```suggestion
      // maybe a problem with endianess can happen?
```

Reply:
I forgot I wrote this and was like "why did I write this" :laughing: